opengrok-indexer/src/main/java/org/opengrok/indexer/web/messages/MessagesUtils.java

```diff
@@ -0,0 +1,245 @@
+                    out.write(": ");
+                    out.write(m.getMessage().getText());
+                    out.write("</li>");
+                }
+                out.write("</ul>");
+            } catch (IOException ex) {
+                LOGGER.log(Level.WARNING,
+                        "An error occurred for a group of messages", ex);
+            }
+        }
+    }
+
+    /**
+     * Print set of tagged messages into JSON object.
+     * TOOD should this be TaggedMessagesContainer.toJSON() ?
+     *
+     * @return JSON string or null
+     */
+    @SuppressWarnings("unchecked")
+    private static String taggedMessagesToJson(Set<TaggedMessagesContainer> messages) {
+        if (messages.isEmpty()) {
+            return emptyJSON;
+        }
+
+        ObjectMapper mapper = new ObjectMapper();
+
+        try {
+            return mapper.writeValueAsString(messages);
+        } catch (JsonProcessingException e) {
+            LOGGER.log(Level.WARNING, "failed to encode {} to JSON", messages);
+            return null;
+        }
+    }
+
+    /**
+     * Print messages for given tags into JSON array
+     *
+     * @param tags list of tags
+     * @return JSON array of the messages (the same as the parameter)
+     */
+    @SuppressWarnings("unchecked")
+    public static String messagesToJson(String... tags) {
+        Set<TaggedMessagesContainer> messages = new HashSet<>();
+
+        for (String tag : tags) {
+            SortedSet<MessagesContainer.AcceptedMessage> messagesWithTag = RuntimeEnvironment.getInstance().getMessages(tag);
+            if (messagesWithTag.isEmpty()) {
+                continue;
+            }
+
+            TaggedMessagesContainer container = new TaggedMessagesContainer(tag, messagesWithTag);
+            messages.add(container);
+        }
+
+        return taggedMessagesToJson(messages);
+    }
+
+    /**
+     * Print messages for given tags into JSON array
+     *
+     * @param tags list of tags
+     * @return json array of the messages
+     * @see #messagesToJson(String...)
+     */
+    private static String messagesToJson(List<String> tags) {
+        String[] array = new String[tags.size()];
+
+        return messagesToJson(tags.toArray(array));
+    }
+
+    /**
+     * Print messages for given project into JSON. These messages are
+     * tagged by project description or tagged by any of the project's group name.
+     *
+     * @param project the project
+     * @param additionalTags additional list of tags
+     * @return JSON string
+     * @see #messagesToJson(String...)
+     */
+    public static String messagesToJson(Project project, String... additionalTags) {
+        if (project == null) {
+            return emptyJSON;
+        }
+
+        List<String> tags = new ArrayList<>();
+        tags.addAll(Arrays.asList(additionalTags));
+        tags.add(project.getName());
+        project.getGroups().stream().forEach((Group t) -> {
```

Comment:
Maybe rewrite to
```java
project.getGroups().stream().map(Group::getName).forEach(tags::add);
```
?

Hopefully I wrote it correctly (no IDE suggestions :D ).

Reply:
Yeah, it's strange combination of 2 styles. In fact, IDEA offers to replace this with loop:
```java
        for (Group t : project.getGroups()) {
            tags.add(t.getName());
        }
```
but I guess these days stream API one liner is more fashionable.

Reply:
then go ahead, I wrote it when I was new in the streaming api